src/Tuner.cpp

```diff
@@ -0,0 +1,424 @@
+    }
+    if (!IsMultiple(p["MWG"], p["MDIMA"]*p["VWM"])) {
+        return false;
+    }
+    if (!IsMultiple(p["NWG"], p["NDIMB"]*p["VWN"])) {
+        return false;
+    }
+    if (!IsMultiple(p["KWG"], p["MDIMC"]*p["NDIMC"]/p["MDIMA"])) {
+        return false;
+    }
+    if (!IsMultiple(p["KWG"], p["MDIMC"]*p["NDIMC"]/p["NDIMB"])) {
+        return false;
+    }
+    if (p["MDIMC"] != p["MDIMA"]) {
+        return false;
+    }
+    if (p["NDIMC"] != p["NDIMB"]) {
+        return false;
+    }
+    if (p["SA"] != p["SB"]) {
+        return false;
+    }
+    return true;
+}
+
+
+Parameters Tuner::get_parameters_by_int(const std::vector<Configurations> opts, const int n) {
+
+        Parameters param;
+        std::vector<size_t> choices(opts.size());
+
+        auto cfgs = 1;
+        for (auto c = size_t{0}; c < opts.size(); c++) {
+            choices[c] = opts[c].second.size();
+            cfgs *= choices[c];
+        }
+        int j = n;
+
+        for (auto c = size_t{0}; c < opts.size(); c++) {
+            auto o = opts[c];
+            auto s = o.first;
+            auto v = o.second[j % choices[c]];
+            j /= choices[c];
+            param[s] = v;
+        }
+
+        return param;
+}
+
+std::string Tuner::parameters_to_string(const Parameters p) {
+    std::string s;
+    for (auto const& x : p) {
+        s += " -D" + x.first + "=" + std::to_string(x.second);
+    }
+    return s;
+}
+
+size_t next_power_of_two(const size_t x) {
+    return 2 << (size_t)(std::ceil(std::log2(x)) - 1);
+}
+
+static void sgemm_generate_data(std::vector<float> &x, const int m, const int n,
+        const int batch_size, const int m_ceil, const int n_ceil) {
+    for (auto batch = 0; batch < batch_size; batch++) {
+        for (auto i = 0; i < n_ceil; i++) {
+            if (i < n) {
+                for (auto j = 0; j < m; j++) {
+                    x[batch*n_ceil*m_ceil + i*m_ceil + j] = 0.01f*(((i ^ j) + batch - 50) % 100);
+                }
+                for (auto j = m; j < m_ceil; j++) {
+                    x[batch*n_ceil*m_ceil + i*m_ceil + j] = 0.0f;
+                }
+            } else {
+                for (auto j = 0; j < m_ceil; j++) {
+                    x[batch*n_ceil*m_ceil + i*m_ceil + j] = 0.0f;
+                }
+            }
+        }
+    }
+}
+
+static float compare_ref(std::vector<float> &x, std::vector<float> &ref,
+        const int m, const int n, const int batch_size) {
+
+    auto sum = 0.0f;
+    for (auto batch = 0; batch < batch_size; batch++) {
+        for (auto i = 0; i < n; i++) {
+            for (auto j = 0; j < m; j++) {
+                auto r = ref[batch*n*m + i*m + j];
+                auto y = x[batch*n*m + j*n + i];
+
+                sum += (r - y) * (r - y);
+            }
+        }
+    }
+    return sum/(m*n);
+}
+
+std::string Tuner::tune_sgemm(const int m, const int n, const int k, const int batch_size, const int runs) {
+
+    std::vector<Configurations> opts = {
```

Comment:
Ah, the how much tuning dilemma. I think you already pruned the possibilities from CLBlast here, right?

I'd propose this:
a) Add an --full-tuning or --exhaustive-tuning flag, which tries all sensible combinations (including the ones already pruned here).
b) Run it on a few systems you have access to. Include the results in the source.
c) By default, we only test the exact combinations from (b).
d) If exhaustive-tuning is requested, and we get an improvement of more than 5% over something from (b), mark that in the tuned output.
e) Ask people to send in the output from (d). Add to source if >5% better.

This should allow getting to within say 5% of optimal performance with maybe 20-30 attempts, covering all common GPU. People who want the last 5% can use the flag.

Reply:
CLBlast has two tuning methods for sgemm. Reduced set of arguments that are tried exhaustively and full set that are tried randomly. This is the reduced set.

I agree that your changes would be good, but I don't have the motivation to implement them. I just wanted to learn OpenCL when I started this pull request and this change goes too far. I'll leave the tuner improvements for someone else.

Reply:
The implement tuner did not consume too much time, so is there any reason for configurations not to remain as is?

Reply:
It doesn't take too much time because it only tries a limited subset, leaving some performance on the table. I'm thinking of ways to recover that without making the first startup take 15 mins for everyone.

Reply:
The exhaustive tuner could likely also be much faster by making it use a genetic algorithm, or CEM or whatever optimization method, rather than blindly trying all combinations. But realistically there's maybe 20 common models of GPU, so if you have the tuning for most of those, it could finish near instantly.